### PR TITLE
refactor image-manipulation logic

### PR DESCRIPTION
**Pull request by wjordan**
This PR consolidates _triplicated_ image-manipulation code contained in `shared_resources.rb`, Pegasus `router.rb` and Pegasus/code.org `images/splat.erb` into a single helper method in `lib/cdo/pegasus/graphics#process_image`.

Also fixes many bugs and inconsistencies between the diverging implementations, and adds a test suite with good coverage across all features and edge-cases of the API.

This is a followup refactoring to #5168 and #3725.

### CODE REVIEW

**Comment by philbogle** [comment]
This seems like it can wait past November 9th?

**Comment by wjordan** [comment]
Yes, definitely past Novermber 9th for this.

**Comment by wjordan** [comment]
@breville can you review? I think these image-transformation routes are used pretty heavily in Pegasus (though the behavior should remain the same).

File: lib/cdo/pegasus/graphics.rb [hunk 1]
```diff
@@ -2,20 +2,100 @@
 
 MAX_DIMENSION = 2880
 
-def load_manipulated_image(path, mode, width, height = nil)
+def load_manipulated_image(path, mode, width, height, scale = nil)
```
Comment (breville) [inline lib/cdo/pegasus/graphics.rb:5]: do we still need height = nil?

File: lib/cdo/pegasus/graphics.rb [hunk 2]
```diff
@@ -2,20 +2,100 @@
...
-    Magick::Image.read(path).first.resize_to_fit(width, height)
-  when :resize
-    height ||= width
-    Magick::Image.read(path).first.resize(width, height)
+    when :fill
+      image.resize_to_fill(width, height)
+    when :fit
+      image.resize_to_fit(width, height)
+    when :resize
+      height ||= width
```
Comment (breville) [inline lib/cdo/pegasus/graphics.rb]: did we already deal with this case above?

File: lib/cdo/pegasus/graphics.rb (hunk 1 above)
Comment (wjordan) [inline lib/cdo/pegasus/graphics.rb:5]: I checked that no existing callers were making use of the default `height` parameter, so i don't think we still need it

File: lib/cdo/pegasus/graphics.rb (hunk 2 above)
Comment (wjordan) [inline lib/cdo/pegasus/graphics.rb]: ah, good catch, yes this line can be removed now that it was added above.

**Comment by breville** [comment]
Looks good, and nice testing.  I tested in on a few pages, including homepage, and things appear to be working as expected.

**Comment by wjordan** [comment]
note-to-self:

this refactor fixed an issue where certain image-processing paths were actually being processed by an older code-path (`splat.erb`) which did not correctly set the image caching headers (it used shorter `document` cache lifetimes, rather than the longer `image` ones). This in turn will affect how often the image-processing code will be invoked on production, so I will be sure to monitor the Pegasus memory profile following our next production deploy to see if this change to cache lifetimes has any direct impact on memory usage.
